core.ts

```diff
@@ -0,0 +1,1041 @@
+
+			// precompile the URL string to improve performance
+			const compiledUrl = this.compile(params)
+			if (utils.isString(params)) {
+				params = compiledUrl
+			} else {
+				params.url = compiledUrl
+				pollInterval = params.pollInterval
+			}
+
+			const requestFn = () => {
+				return this.request(params, { method: 'GET' }).then((data: {d: any[]}) => {
+					if (!utils.isObject(data)) {
+						throw new Error(`Response was not a JSON object: '${typeof data}'`)
+					}
+					if (data.d == null) {
+						throw new Error("Invalid response received, the 'd' property is missing.")
+					}
+					if (singular) {
+						if (data.d.length > 1) {
+							throw new Error('Returned multiple results when only one was expected.')
+						}
+						return data.d[0]
+					}
+					return data.d
+				}) as PromiseResult
+			}
+
+			return new Poll(requestFn, pollInterval)
+		}
+
+		put(params: PinejsClientCoreFactory.Params) {
+			return this.request(params, { method: 'PUT' })
+		}
+
+		patch(params: PinejsClientCoreFactory.Params) {
+			return this.request(params, { method: 'PATCH' })
+		}
+
+		post(params: PinejsClientCoreFactory.Params) {
+			return this.request(params, { method: 'POST' })
+		}
+
+		delete(params: PinejsClientCoreFactory.Params) {
+			return this.request(params, { method: 'DELETE' })
+		}
+
+		compile(params: PinejsClientCoreFactory.Params) {
+			if (utils.isString(params)) {
+				return params
+			} else if (params.url != null) {
+				return params.url
+			} else {
+				if (params.resource == null) {
+					throw new Error('Either the url or resource must be specified.')
+				}
+				let url = params.resource
+
+				if (params.hasOwnProperty('id')) {
+					if (params.id == null) {
+						throw new Error('If the id property is set it must be non-null')
+					}
+					url += `(${escapeValue(params.id)})`
+				}
+
+				let queryOptions: string[] = []
+				if (params.options != null) {
+					queryOptions = mapObj(params.options, (value, option) => {
+						if(option[0] === '$' && !isValidOption(option)) {
+							throw new Error(`Unknown odata option '${option}'`)
+						}
+						return buildOption(option, value)
+					})
+				}
+				if ((params as any).customOptions != null) {
+					throw new Error('`customOptions` has been removed, use `options` instead.')
+				}
+				if (queryOptions.length > 0) {
+					url += '?' + queryOptions.join('&')
+				}
+				return url
+			}
+		}
+
+		request(params: PinejsClientCoreFactory.Params, overrides: { method?: PinejsClientCoreFactory.ODataMethod } = {}): PromiseObj {
+			try {
+				let method: PinejsClientCoreFactory.ParamsObj['method']
+				let body: PinejsClientCoreFactory.ParamsObj['body']
+				let passthrough: PinejsClientCoreFactory.ParamsObj['passthrough'] = {}
+				let apiPrefix: PinejsClientCoreFactory.ParamsObj['apiPrefix']
+
+
+				if (utils.isString(params)) {
+					method = 'GET'
+				} else {
+					({ method, body, passthrough = {}, apiPrefix } = params)
```

Comment:
Yes, because otherwise the `{}` are seen as normal braces around a code block rather than for destructuring - you don't need it for `const/let/var { ... } = ...` because the `const/let/var` can only be followed by a var name or destructuring operator